qiskit/result/utils.py

```diff
@@ -52,63 +54,84 @@ def marginal_counts(result, indices=None, inplace=False):
                 new_counts_hex[_bin_to_hex(k)] = v
             experiment_result.data.counts = new_counts_hex
             experiment_result.header.memory_slots = len(indices)
+            csize = experiment_result.header.creg_sizes
+            experiment_result.header.creg_sizes = _adjust_creg_sizes(csize, indices)
+        return result
     else:
-        result = _marginalize(result, indices)
-
-    return result
-
-
-def count_keys(num_clbits):
-    """Return ordered count keys."""
-    return [bin(j)[2:].zfill(num_clbits) for j in range(2 ** num_clbits)]
+        marg_counts = _marginalize(result, indices)
+        if format_marginal and indices is not None:
+            marg_counts = _format_marginal(result, marg_counts, indices)
+        return marg_counts
+
+
+def _adjust_creg_sizes(creg_sizes, indices):
+    """Helper to reduce creg_sizes to match indices"""
+
+    # Zero out creg_sizes list
+    new_creg_sizes = [[creg[0], 0] for creg in creg_sizes]
+    indices_sort = sorted(indices)
+
+    # Get creg num values and then convert to the cumulative last index per creg.
+    # e.g. [2, 1, 3] => [1, 2, 5]
+    creg_nums = [x for _, x in creg_sizes]
+    creg_limits = [sum(creg_nums[0:x:1])-1 for x in range(0, len(creg_nums)+1)][1:]
+
+    # Now iterate over indices and find which creg that index is in.
+    # When found increment the creg size
+    creg_idx = 0
+    for ind in indices_sort:
+        for idx in range(creg_idx, len(creg_limits)):
+            if ind <= creg_limits[idx]:
+                creg_idx = idx
+                new_creg_sizes[idx][1] += 1
+                break
+    # Throw away any cregs with 0 size
+    new_creg_sizes = [creg for creg in new_creg_sizes if creg[1] != 0]
+    return new_creg_sizes
 
 
 def _marginalize(counts, indices=None):
-    # Extract total number of clbits from first count key
-    # We trim the whitespace separating classical registers
-    # and count the number of digits
+    """Get the marginal counts for the given set of indices"""
     num_clbits = len(next(iter(counts)).replace(' ', ''))
 
-    # Check if we do not need to marginalize. In this case we just trim
-    # whitespace from count keys
+    # Check if we do not need to marginalize and if so, trim
+    # whitespace and '_' and return
     if (indices is None) or set(range(num_clbits)) == set(indices):
         ret = {}
         for key, val in counts.items():
-            key = key.replace(' ', '')
+            key = key.replace(' ', '').replace('_', '')
             ret[key] = val
```

Comment:
This pattern repeats a lot. Maybe worthy to make a function (in `counts.py`, I assume) for it?

Reply:
Added a function to both counts.py (3 uses) and utils.py (2 uses).